File: Algorithms/Merge Sort/Merge Sort.cpp
template <class T>
void merge(T arr[], int low, int mid, int high)
{
    T aux[high - low + 1];
    int left = low, right = mid + 1;
    int i = 0;

    while (left <= mid && right <= high)
    {
        if (arr[left] <= arr[right])
            aux[i++] = arr[left++];
        if (arr[left] >= arr[right])
            aux[i++] = arr[right++];
    }

    while (left <= mid)
        aux[i++] = arr[left++];
    while (right <= high)
        aux[i++] = arr[right++];

    for (int j = 0; j < i; j++)
        arr[low + j] = aux[j];
}

template <class T>
void merge_sort(T arr[], int low, int high)
{
    if (high <= low)
        return;

    int mid = (high + low) / 2;
    merge_sort<T>(arr, low, mid);
    merge_sort<T>(arr, mid + 1, high);

    merge<T>(arr, low, mid, high);
}

File: Algorithms/Sorting/Merge Sort/Merge Sort.py
def merge(arr, low, mid, high):
    aux = [0] * (high - low + 1)
    left, right, i = low, mid + 1, 0

    while left <= mid and right <= high:
        if arr[left] <= arr[right]:
            aux[i] = arr[left]
            left += 1
            i += 1
        if arr[left] >= arr[right]:
            aux[i] = arr[right]
            right += 1
            i += 1

    while left <= mid:
        aux[i] = arr[left]
        left += 1
        i += 1
    while right <= high:
        aux[i] = arr[right]
        right += 1
        i += 1

    arr[low : high + 1] = aux


def merge_sort(arr, low, high):
    if low >= high:
        return

    mid = (high + low) // 2
    merge_sort(arr, low, mid)
    merge_sort(arr, mid + 1, high)

    merge(arr, low, mid, high)
